### ISSUE TITLE
ZFS unmount fails in various tests

### ISSUE THREAD

**Issue by ZackaryWelch** [zfsonfreebsd/ZoF#35]
In the test snapshot/rollback_003, a command at the end of the test to unmount and then remount drives in the testpool fails as follows:
```
cannot unmount 'testpool/testctr/testfs1': Invalid argument cannot unmount 'testpool/testfs': Invalid argument cannot unmount 'testpool/testfile-snapshot': Invalid argument cannot unmount 'testpool': Invalid argument
ERROR: zfs unmount -a exited 1
```
The full segment of the script is:
```
export __ZFS_POOL_RESTRICT="$TESTPOOL"
log_must zfs unmount -a
log_must zfs mount -a
unset __ZFS_POOL_RESTRICT
```
This error might be due to the export variable failing to have any significance.

Edit:
This error can also be seen in cli_root/zfs_unmount/zfs_unmount_009, zfs_unmount_all_001 and cli_root/zfs_mount/zfs_mount_009, zfs_mount_all_001, zfs_mount_all_fail, and zfs_mount_all_mountpoints as:
```
cannot unmount 'testpool/testfs1': Invalid argument cannot unmount 'testpool/testfs': Invalid argument cannot unmount 'testpool': Invalid argument
ERROR: zfs unmount -a exited 1
```

**Comment by ZackaryWelch** [comment]
Setup
in rollback_003:
```
(in setup)
zpool create -f testpool /dev/md0 /dev/md1 /dev/md2
mkdir -p /var/tmp/testdir
zfs create testpool/testfs
zfs set mountpoint=/var/tmp/testdir testpool/testfs
zfs snapshot testpool@testsnap
mkdir -p /tests/testdir1
zfs create testpool/testctr
zfs set canmount=off testpool/testctr
zfs create testpool/testctr/testfs1
zfs set mountpoint=/tests/testdri1 testpool/testctr/testfs1
zfs create -V 150m testpool/testvol

(in rollback)
zfs snapshot testpool@testsnap
zfs rollback testpool@testsnap
zfs snapshot testpool@testsnap (should fail)
touch /testpool/testfile-snapshot
zfs rollback testpool@testsnap
zfs create /testpool/testfile-snapshot
zfs rollback testpool@testsnap
zfs snapshot testpool/testfile-snapshot@testsnap
zfs snapshot testpool@testsnap.1

export __ZFS_POOL_RESTRICT="testpool"
zfs unmount -a
zfs mount -a
unset __ZFS_POOL_RESTRICT
```
Might be missing a little bit of setup, but this seems it. The only relevant part might be the last 4 lines.

**Comment by mattmacy** [comment]
Is this what do you see?

```
beastie3# zfs unmount -a
cannot unmount 'testpool/testfs': Invalid argument
cannot unmount 'testpool/testctr/testfs1': Invalid argument
cannot unmount 'testpool/testfile-snapshot': Invalid argument
cannot unmount 'testpool': Invalid argument
beastie3# zfs mount -a
beastie3# unset __ZFS_POOL_RESTRICT
beastie3# df -h
Filesystem                    Size    Used   Avail Capacity  Mounted on
/dev/ada0p2                    22G     15G    5.2G    75%    /
devfs                         1.0K    1.0K      0B   100%    /dev
testpool                      434M     24K    434M     0%    /testpool
testpool/testfs               434M     24K    434M     0%    /var/tmp/testdir
testpool/testctr/testfs1      434M     24K    434M     0%    /tests/testdri1
testpool/testfile-snapshot    434M     24K    434M     0%    /testpool/testfile-snapshot

```
And what is the behavior on ZFS in FreeBSD?

**Comment by ZackaryWelch** [comment]
Updated my command list. They're no messages when the test runs on regular ZFS.

**Comment by mattmacy** [comment]
fixed by 110595a7d71b42912162dbba3def420d35b443c3
